File: firmware/components/forgesense_protocol/forgesense_protocol.cpp
#include "forgesense_protocol.h"

namespace forgesense {
namespace {

constexpr std::size_t kHeaderSize = 12;
constexpr std::size_t kMlPayloadSize = 14;
constexpr std::size_t kSensorPayloadSize = 8;

std::uint16_t read_u16(const std::uint8_t* p) {
    return static_cast<std::uint16_t>(p[0]) |
           (static_cast<std::uint16_t>(p[1]) << 8);
}

std::int16_t read_i16(const std::uint8_t* p) {
    return static_cast<std::int16_t>(read_u16(p));
}

std::uint32_t read_u32(const std::uint8_t* p) {
    return static_cast<std::uint32_t>(p[0]) |
           (static_cast<std::uint32_t>(p[1]) << 8) |
           (static_cast<std::uint32_t>(p[2]) << 16) |
           (static_cast<std::uint32_t>(p[3]) << 24);
}

void write_u16(std::uint8_t* p, std::uint16_t value) {
    p[0] = static_cast<std::uint8_t>(value & 0xFF);
    p[1] = static_cast<std::uint8_t>((value >> 8) & 0xFF);
}

void write_u32(std::uint8_t* p, std::uint32_t value) {
    p[0] = static_cast<std::uint8_t>(value & 0xFF);
    p[1] = static_cast<std::uint8_t>((value >> 8) & 0xFF);
    p[2] = static_cast<std::uint8_t>((value >> 16) & 0xFF);
    p[3] = static_cast<std::uint8_t>((value >> 24) & 0xFF);
}

bool health_valid(std::uint8_t value) {
    return value <= static_cast<std::uint8_t>(HealthClass::Abstain);
}

ParseStatus validate_common(
    const std::uint8_t* data,
    std::size_t size,
    std::uint8_t expected_type,
    std::uint16_t expected_payload_size) {
    if (data == nullptr || size < kHeaderSize + 2) {
        return ParseStatus::Truncated;
    }
    if (data[0] != 0xA5 || data[1] != 0x5A) {
        return ParseStatus::BadSof;
    }
    if (data[2] != kProtocolVersion) {
        return ParseStatus::BadVersion;
    }
    if (data[3] != expected_type) {
        return ParseStatus::BadType;
    }
    const auto payload_size = read_u16(data + 6);
    const std::size_t expected_frame_size =
        kHeaderSize + expected_payload_size + 2;
    if (payload_size != expected_payload_size || size != expected_frame_size) {
        return ParseStatus::BadLength;
    }
    const auto received_crc = read_u16(data + size - 2);
    const auto calculated_crc = crc16_ccitt(data + 2, size - 4);
    if (received_crc != calculated_crc) {
        return ParseStatus::BadCrc;
    }
    return ParseStatus::Ok;
}

}  // namespace

std::uint16_t crc16_ccitt(const std::uint8_t* data, std::size_t size) {
    std::uint16_t crc = 0xFFFF;
    for (std::size_t i = 0; i < size; ++i) {
        crc ^= static_cast<std::uint16_t>(data[i]) << 8;
        for (int bit = 0; bit < 8; ++bit) {
            crc = (crc & 0x8000)
                ? static_cast<std::uint16_t>((crc << 1) ^ 0x1021)
                : static_cast<std::uint16_t>(crc << 1);
        }
    }
    return crc;
}

ParseStatus parse_ml_frame(
    const std::uint8_t* data,
    std::size_t size,
    ParsedMlFrame& out) {
    const auto common =
        validate_common(data, size, kMessageMlObservation, kMlPayloadSize);
    if (common != ParseStatus::Ok) {
        return common;
    }

    const std::uint8_t* payload = data + kHeaderSize;
    if (!health_valid(payload[10])) {
        return ParseStatus::BadHealthClass;
    }

    const auto anomaly_q15 = read_u16(payload + 8);
    if (anomaly_q15 > 32767) {
        return ParseStatus::BadRange;
    }

    out.sequence = read_u16(data + 4);
    out.timestamp_ms = read_u32(data + 8);
    out.observation.model_id = read_u16(payload + 0);
    out.observation.model_version = read_u16(payload + 2);
    out.observation.feature_schema_version = read_u16(payload + 4);
    out.observation.flags = read_u16(payload + 6);
    out.observation.anomaly_q15 = anomaly_q15;
    out.observation.health_class = static_cast<HealthClass>(payload[10]);
    out.observation.confidence_q8 = payload[11];
    out.observation.inference_age_ms = read_u16(payload + 12);
    return ParseStatus::Ok;
}

ParseStatus parse_sensor_frame(
    const std::uint8_t* data,
    std::size_t size,
    ParsedSensorFrame& out) {
    const auto common =
        validate_common(data, size, kMessageSensorSnapshot, kSensorPayloadSize);
    if (common != ParseStatus::Ok) {
        return common;
    }

    const std::uint8_t* payload = data + kHeaderSize;
    out.sequence = read_u16(data + 4);
    out.timestamp_ms = read_u32(data + 8);
    out.snapshot.temperature_deci_c = read_i16(payload + 0);
    out.snapshot.vibration_milli_g = read_u16(payload + 2);
    out.snapshot.current_milli_a = read_u16(payload + 4);
    out.snapshot.flags = read_u16(payload + 6);
    return ParseStatus::Ok;
}

bool encode_ml_frame(
    const MlObservation& observation,
    std::uint16_t sequence,
    std::uint32_t timestamp_ms,
    std::array<std::uint8_t, kMlFrameSize>& out) {
    if (observation.anomaly_q15 > 32767 ||
        !health_valid(static_cast<std::uint8_t>(observation.health_class))) {
        return false;
    }

    out.fill(0);
    out[0] = 0xA5;
    out[1] = 0x5A;
    out[2] = kProtocolVersion;
    out[3] = kMessageMlObservation;
    write_u16(out.data() + 4, sequence);
    write_u16(out.data() + 6, static_cast<std::uint16_t>(kMlPayloadSize));
    write_u32(out.data() + 8, timestamp_ms);

    std::uint8_t* payload = out.data() + kHeaderSize;
    write_u16(payload + 0, observation.model_id);
    write_u16(payload + 2, observation.model_version);
    write_u16(payload + 4, observation.feature_schema_version);
    write_u16(payload + 6, observation.flags);
    write_u16(payload + 8, observation.anomaly_q15);
    payload[10] = static_cast<std::uint8_t>(observation.health_class);
    payload[11] = observation.confidence_q8;
    write_u16(payload + 12, observation.inference_age_ms);

    const auto crc = crc16_ccitt(out.data() + 2, kMlFrameSize - 4);
    write_u16(out.data() + kMlFrameSize - 2, crc);
    return true;
}

bool sequence_is_newer(std::uint16_t candidate, std::uint16_t previous) {
    const auto delta = static_cast<std::uint16_t>(candidate - previous);
    return delta != 0 && delta < 0x8000;
}

}  // namespace forgesense

File: protocol/python/forgesense_protocol.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
import struct

SOF = b"\xA5\x5A"
PROTOCOL_VERSION = 1

VALID_OBSERVATION = 0x0001

SENSOR_VALID_TEMPERATURE = 0x0001
SENSOR_VALID_VIBRATION = 0x0002
SENSOR_VALID_CURRENT = 0x0004
SENSOR_ALL_VALID = SENSOR_VALID_TEMPERATURE | SENSOR_VALID_VIBRATION | SENSOR_VALID_CURRENT

STATUS_LOAD_ENABLE = 0x01
STATUS_WARNING_ACTIVE = 0x02
STATUS_FAULT_LATCHED = 0x04
STATUS_OPERATIONAL_READY = 0x08

SAFETY_HARD_WARNING = 0x0001
SAFETY_HARD_CRITICAL = 0x0002
SAFETY_COMM_TIMEOUT = 0x0004
SAFETY_ML_WARNING = 0x0008
SAFETY_ML_CRITICAL = 0x0010
SAFETY_EMERGENCY = 0x0020
SAFETY_SENSORS_VALID = 0x0040
SAFETY_DEVICE_IDENTITY_OK = 0x0080
SAFETY_DEVICE_TRANSPORT_ERROR = 0x0100

_HEADER = struct.Struct("<2sBBHHI")
_ML_PAYLOAD = struct.Struct("<HHHHHBBH")
_SENSOR_PAYLOAD = struct.Struct("<hHHH")
_STATUS_PAYLOAD = struct.Struct("<BBH")
_CRC = struct.Struct("<H")


class ProtocolError(ValueError):
    pass


class MessageType(IntEnum):
    ML_OBSERVATION = 0x10
    SENSOR_SNAPSHOT = 0x11
    HEARTBEAT = 0x20
    STATUS = 0x30
    EVENT = 0x31


class HealthClass(IntEnum):
    NORMAL = 0
    WARNING = 1
    CRITICAL = 2
    ABSTAIN = 3


@dataclass(frozen=True)
class Frame:
    version: int
    message_type: int
    sequence: int
    timestamp_ms: int
    payload: bytes


@dataclass(frozen=True)
class MlObservation:
    model_id: int
    model_version: int
    feature_schema_version: int
    flags: int
    anomaly_score: float
    health_class: HealthClass
    confidence: float
    inference_age_ms: int


@dataclass(frozen=True)
class StatusSnapshotWire:
    state_code: int
    control_flags: int
    safety_flags: int

    @property
    def load_enable(self) -> bool:
        return bool(self.control_flags & STATUS_LOAD_ENABLE)

    @property
    def warning_active(self) -> bool:
        return bool(self.control_flags & STATUS_WARNING_ACTIVE)

    @property
    def fault_latched(self) -> bool:
        return bool(self.control_flags & STATUS_FAULT_LATCHED)

    @property
    def operational_ready(self) -> bool:
        return bool(self.control_flags & STATUS_OPERATIONAL_READY)

    @property
    def device_identity_ok(self) -> bool:
        return bool(self.safety_flags & SAFETY_DEVICE_IDENTITY_OK)

    @property
    def device_transport_error(self) -> bool:
        return bool(self.safety_flags & SAFETY_DEVICE_TRANSPORT_ERROR)


@dataclass(frozen=True)
class SensorSnapshotWire:
    temperature_deci_c: int
    vibration_milli_g: int
    current_milli_a: int
    flags: int = SENSOR_ALL_VALID

    @property
    def all_valid(self) -> bool:
        return (self.flags & SENSOR_ALL_VALID) == SENSOR_ALL_VALID

    @property
    def temperature_c(self) -> float:
        return self.temperature_deci_c / 10.0

    @property
    def vibration_rms_g(self) -> float:
        return self.vibration_milli_g / 1000.0

    @property
    def current_a(self) -> float:
        return self.current_milli_a / 1000.0

    def feature_vector(self) -> tuple[float, float, float]:
        if not self.all_valid:
            raise ValueError("invalid sensor snapshot cannot be converted to an ML feature vector")
        return self.temperature_c, self.vibration_rms_g, self.current_a


def crc16_ccitt(data: bytes, initial: int = 0xFFFF) -> int:
    crc = initial
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) & 0xFFFF if crc & 0x8000 else (crc << 1) & 0xFFFF
    return crc


def encode_frame(
    *,
    message_type: int,
    sequence: int,
    timestamp_ms: int,
    payload: bytes,
    version: int = PROTOCOL_VERSION,
) -> bytes:
    if not (0 <= sequence <= 0xFFFF):
        raise ProtocolError("sequence out of range")
    if not (0 <= timestamp_ms <= 0xFFFFFFFF):
        raise ProtocolError("timestamp out of range")
    if len(payload) > 0xFFFF:
        raise ProtocolError("payload too large")
    header = _HEADER.pack(SOF, version, int(message_type), sequence, len(payload), timestamp_ms)
    body = header[2:] + payload
    return header + payload + _CRC.pack(crc16_ccitt(body))


def decode_frame(data: bytes, *, expected_version: int = PROTOCOL_VERSION) -> Frame:
    minimum = _HEADER.size + _CRC.size
    if len(data) < minimum:
        raise ProtocolError("frame truncated")
    sof, version, message_type, sequence, payload_len, timestamp_ms = _HEADER.unpack_from(data)
    if sof != SOF:
        raise ProtocolError("invalid start marker")
    expected_len = _HEADER.size + payload_len + _CRC.size
    if len(data) != expected_len:
        raise ProtocolError("frame length mismatch")
    if version != expected_version:
        raise ProtocolError("unsupported protocol version")
    payload = data[_HEADER.size : _HEADER.size + payload_len]
    received_crc = _CRC.unpack_from(data, _HEADER.size + payload_len)[0]
    calculated_crc = crc16_ccitt(data[2 : _HEADER.size + payload_len])
    if received_crc != calculated_crc:
        raise ProtocolError("CRC mismatch")
    return Frame(version, message_type, sequence, timestamp_ms, payload)


def encode_ml_observation(observation: MlObservation, *, sequence: int, timestamp_ms: int) -> bytes:
    score_q15 = round(min(max(observation.anomaly_score, 0.0), 1.0) * 32767)
    confidence_q8 = round(min(max(observation.confidence, 0.0), 1.0) * 255)
    if not (0 <= observation.inference_age_ms <= 0xFFFF):
        raise ProtocolError("inference age out of range")
    payload = _ML_PAYLOAD.pack(
        observation.model_id,
        observation.model_version,
        observation.feature_schema_version,
        observation.flags,
        score_q15,
        int(observation.health_class),
        confidence_q8,
        observation.inference_age_ms,
    )
    return encode_frame(
        message_type=MessageType.ML_OBSERVATION,
        sequence=sequence,
        timestamp_ms=timestamp_ms,
        payload=payload,
    )


def decode_ml_observation(frame: Frame) -> MlObservation:
    if frame.message_type != MessageType.ML_OBSERVATION:
        raise ProtocolError("not an ML observation frame")
    if len(frame.payload) != _ML_PAYLOAD.size:
        raise ProtocolError("invalid ML observation payload length")
    model_id, model_version, schema, flags, score_q15, health, confidence_q8, age_ms = _ML_PAYLOAD.unpack(frame.payload)
    if score_q15 > 32767:
        raise ProtocolError("invalid Q15 anomaly score")
    try:
        health_class = HealthClass(health)
    except ValueError as exc:
        raise ProtocolError("invalid health class") from exc
    return MlObservation(
        model_id=model_id,
        model_version=model_version,
        feature_schema_version=schema,
        flags=flags,
        anomaly_score=score_q15 / 32767.0,
        health_class=health_class,
        confidence=confidence_q8 / 255.0,
        inference_age_ms=age_ms,
    )


def encode_sensor_snapshot(
    snapshot: SensorSnapshotWire,
    *,
    sequence: int,
    timestamp_ms: int,
) -> bytes:
    if not (-32768 <= snapshot.temperature_deci_c <= 32767):
        raise ProtocolError("temperature out of range")
    if not (0 <= snapshot.vibration_milli_g <= 0xFFFF):
        raise ProtocolError("vibration out of range")
    if not (0 <= snapshot.current_milli_a <= 0xFFFF):
        raise ProtocolError("current out of range")
    if not (0 <= snapshot.flags <= 0xFFFF):
        raise ProtocolError("sensor flags out of range")
    payload = _SENSOR_PAYLOAD.pack(
        snapshot.temperature_deci_c,
        snapshot.vibration_milli_g,
        snapshot.current_milli_a,
        snapshot.flags,
    )
    return encode_frame(
        message_type=MessageType.SENSOR_SNAPSHOT,
        sequence=sequence,
        timestamp_ms=timestamp_ms,
        payload=payload,
    )


def decode_sensor_snapshot(frame: Frame) -> SensorSnapshotWire:
    if frame.message_type != MessageType.SENSOR_SNAPSHOT:
        raise ProtocolError("not a sensor snapshot frame")
    if len(frame.payload) != _SENSOR_PAYLOAD.size:
        raise ProtocolError("invalid sensor snapshot payload length")
    temperature, vibration, current, flags = _SENSOR_PAYLOAD.unpack(frame.payload)
    return SensorSnapshotWire(
        temperature_deci_c=temperature,
        vibration_milli_g=vibration,
        current_milli_a=current,
        flags=flags,
    )


def encode_status_snapshot(
    status: StatusSnapshotWire,
    *,
    sequence: int,
    timestamp_ms: int,
) -> bytes:
    if not (0 <= status.state_code <= 5):
        raise ProtocolError("status state code out of range")
    if not (0 <= status.control_flags <= 0xFF):
        raise ProtocolError("status control flags out of range")
    if not (0 <= status.safety_flags <= 0xFFFF):
        raise ProtocolError("status safety flags out of range")
    payload = _STATUS_PAYLOAD.pack(
        status.state_code, status.control_flags, status.safety_flags
    )
    return encode_frame(
        message_type=MessageType.STATUS,
        sequence=sequence,
        timestamp_ms=timestamp_ms,
        payload=payload,
    )


def decode_status_snapshot(frame: Frame) -> StatusSnapshotWire:
    if frame.message_type != MessageType.STATUS:
        raise ProtocolError("not a status snapshot frame")
    if len(frame.payload) != _STATUS_PAYLOAD.size:
        raise ProtocolError("invalid status payload length")
    state_code, control_flags, safety_flags = _STATUS_PAYLOAD.unpack(frame.payload)
    if state_code > 5:
        raise ProtocolError("invalid status state code")
    return StatusSnapshotWire(state_code, control_flags, safety_flags)


def sequence_is_newer(candidate: int, previous: int) -> bool:
    delta = (candidate - previous) & 0xFFFF
    return 0 < delta < 0x8000


@dataclass
class FreshnessGate:
    expected_model_id: int
    expected_model_version: int
    expected_feature_schema_version: int
    max_inference_age_ms: int = 1500
    last_sequence: int | None = None

    def accept(self, frame: Frame, observation: MlObservation) -> tuple[bool, str]:
        if not observation.flags & VALID_OBSERVATION:
            return False, "observation-invalid"
        if observation.model_id != self.expected_model_id:
            return False, "model-id"
        if observation.model_version != self.expected_model_version:
            return False, "model-version"
        if observation.feature_schema_version != self.expected_feature_schema_version:
            return False, "feature-schema"
        if observation.inference_age_ms > self.max_inference_age_ms:
            return False, "stale-inference"
        if self.last_sequence is not None and not sequence_is_newer(frame.sequence, self.last_sequence):
            return False, "replay-or-duplicate"
        self.last_sequence = frame.sequence
        return True, "accepted"
